kissim/comparison/tree.py

```diff
@@ -0,0 +1,309 @@
+    if annotation_path:
+        _to_kinase_annotation(distance_matrix, annotation_path)
+
+    return tree, node_means
+
+
+def _to_newick(tree, node_means, distance_matrix, outputfile):
+    """
+    Save Newick tree to file.
+
+    Parameters
+    ----------
+    tree : scipy.cluster.hierarchy.ClusterNode
+        Cluster node.
+    node_means : dict of int: float
+        Mean similarity (value) for each node index (key).
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    outputfile : str or pathlib.Path
+        Path to kinase tree file (TREE file) in Newick format.
+    """
+
+    outputfile = Path(outputfile)
+
+    logger.info(f"Writing resulting tree to {outputfile}")
+    newick = ""
+    newick = _get_newick(tree, newick, tree.dist, list(distance_matrix), node_means)
+    with open(outputfile, "w") as f:
+        f.write(newick)
+
+
+def _to_kinase_annotation(distance_matrix, outputfile):
+    """
+    Save kinase annotations to file used for FigTree.
+
+    Parameters
+    ----------
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    outputfile : str or pathlib.Path
+        Path to kinase annotation file (CSV file) in FigTree format.
+    """
+
+    outputfile = Path(outputfile)
+
+    logger.info(f"Writing resulting kinase annotation to {outputfile}")
+
+    # Get kinase names from matrix
+    kinase_names = distance_matrix.columns.to_list()
+
+    # Query KLIFS for kinase details
+    klifs_session = klifs.setup_remote()
+    kinases = klifs_session.kinases.by_kinase_name(kinase_names)
+    kinases = kinases[kinases["species.klifs"] == "Human"]
+    kinases = kinases[["kinase.klifs_name", "kinase.family", "kinase.group"]]
+
+    # Save to file
+    kinases.to_csv(outputfile, sep="\t", index=False)
+
+
+def _get_node_means(tree, distance_matrix, node_means, similarity=True):
+    """
+    Calculating and assign the mean similarity (default) or mean distance for tree branches.
+
+    Parameters
+    ----------
+    tree : scipy.cluster.hierarchy.ClusterNode
+        Clustering tree.
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    node_means : dict of int: float
+        Mean distance (value) for each node index (key).
+    similarity : bool
+        If `True`, convert distance matrix into similarity matrix using
+        `1 - distance_matrix / distance_matrix.max().max()`, else use distance matrix directly.
+
+    Returns
+    -------
+    node_means : dict of int: float
+        Mean similarities (value) for each node index (key).
+        Note: The return value is the populated input `node_means` object.
+    """
+
+    # Calculate mean similarities or mean distances?
+    if similarity:
+        matrix = (1 - distance_matrix / distance_matrix.max().max()).copy()
+        leaf_mean = 1
+    else:
+        matrix = distance_matrix.copy()
+        leaf_mean = 0
```

Comment:
Here we now convert the distance matrix into a similarity matrix (default) or not.

Reply:
Seems like a good solution.
However, I would do the conversion to before calling this function. Since it's a recursive function, it's best to keep it as light as possible and to reduce the overhead.

Reply:
> However, I would do the conversion to before calling this function. Since it's a recursive function, it's best to keep it as light as possible and to reduce the overhead.

Very good point, adapted this accordingly.

> One other thing to keep in mind: not having a fixed distance range will prevent the direct comparison of trees as in those cases the measure will depend on the input. This is at least the case for the current setup using the similarity. If the distance is not influenced by this, then it might be a reason to revert back and use the distance by default instead of the similarity.

Another good point - true, mean similarities could not be compared for different datasets. As you suggested I set the mean distance as default.